### PR TITLE
feat: configure WDK configuration via parsing Cargo manifest metadata

### LINKED ISSUE #82: Support WDK Configuration in crates.io-distributed crate
Currently, the crate that's distributed via crates.io has several things hardcoded in the configuration. Some examples of this include [the WdfFunctions_01033 functional table reference](https://github.com/microsoft/windows-drivers-rs/blob/9202e28d93232824b5469ac8f64f3980220024f5/crates/wdk-sys/src/lib.rs#L63C9-L63C63), and the [WdfMinimumVersionDesired symbol](https://github.com/microsoft/windows-drivers-rs/blob/9202e28d93232824b5469ac8f64f3980220024f5/crates/wdk-sys/src/wdf.rs#L23). The `wdk-build::Config` used in `wdk-sys` also defaults to [KMDF 1.33](https://github.com/microsoft/windows-drivers-rs/blob/9202e28d93232824b5469ac8f64f3980220024f5/crates/wdk-build/src/lib.rs#L618).

While the repo supports different WDF versions, different driver models (WDM, KMDF, UMDF), etc, it is currently cumbersome to use anything that is not KMDF v1.33. Our goal should be to expose allow the user to be able to use the crate off crates.io and be able to configure all of the common configuration options of the WDK. 

A possible way to do this would be to have mutually-exclusive [cargo-features](https://doc.rust-lang.org/cargo/reference/features.html#features) for `wdk-sys` and `wdk` that configure what bindings are created by the crate.
There are some [caveats](https://doc.rust-lang.org/cargo/reference/features.html#mutually-exclusive-features) for using mutually exclusive features, but it might be the best way forward that's currently available. There is extensive discussion about better ways to accomplish this in the future [here](https://internals.rust-lang.org/t/pre-rfc-mutually-excusive-global-features/19618/11) and [here](https://github.com/rust-lang/cargo/issues/2980), but we must make do with what is available right now. 

Another possible approach is to use environment variables to control the various features. Env variables can only have one value at any given time, so they are already mutually exclusive. They could be set via normal environment variable in shell, via [`env` section in a `config.toml`](https://doc.rust-lang.org/nightly/cargo/reference/config.html#env), or in [any of the numerous ways cargo-make provides](https://github.com/sagiegurari/cargo-make?tab=readme-ov-file#usage-env). This does harm discoverability of options, and there are some issues with some ways to configure env variables in Cargo right now (ex. https://github.com/rust-lang/cargo/issues/10358. This one in particular seems like it will not be resolved for a while).

Some other thoughts:
* How will this support multiple drivers in the same workspace? 
  * If we go the path of mutually-exclusive features, [workspace dependency resolution](https://doc.rust-lang.org/cargo/reference/resolver.html#features) would make it so that each cargo workspace can only contain one combination of  mutually-exclusive features. Its possible to get multiple drivers in the same file tree to have different combinations of mutually-exclusive features, by having multiple subfolders that are each a cargo workspace and orchestrating the entire build via cargo-make
  * The environment variable approach would have the same caveats as features.
* Cargo currently has a 300 feature limit due to an [incident](https://blog.rust-lang.org/2023/10/26/broken-badges-and-23k-keywords.html) where the icondata crate (v0.1.0 and earlier) used a [build crate to generate over 23000 mutually exclusive features](https://github.com/Carlosted/icondata/blob/v0.1.0/build/README.md). I don't think we will get anywhere near 300 features though.

### PR DESCRIPTION
**Pull request by wmmc88**
This PR enables configuring the WDK Metadata via `[metadata.wdk]` tags in Cargo manifests.

As a consequence of that new feature, several other changes were required:
* Added minimal Sample WDM and UMDF drivers
* Make all driver packaging tooling (ex. signing, cat generation, etc) support WDM and UMDF
* Serde Serializer added to export WDKMetadata configuration to a structured format
* Emit that structured format so that downstream code can `cfg` based off of certain WDK configuration parameters
* A `wdf.c` file is compiled and linked in to provide the `__declspec(selectany)` that are not generated by bindgen
* Have several templates that generate rust code in `wdk-sys`'s build.rs that is otherwise not possible to generate conditionally (ex. call_unsafe_wdf_function_binding definition that points to `OUT_DIR`, `WDFFUNCTIONS` symbol name that depends on configuration, `WDF_FUNCTION_TABLE` definition whose size depends on configuration, etc.)
* added UM support 
* remove local `generated_bindings` copy of bindings
* exporting via `links` keys in cargo manifests is deprecated
* switch to cargo-machete for unused dep detection (required since most code is not cfg-ed out)
* remove wdk-macros running of `cargo check` to obtain out dir (and replace with wdk_sys passing the outdir to the procmacro)

resolves #82

### CODE REVIEW

**Comment by NateD-MSFT** [comment]
Detailed review is in progress, but quick design question: can you help me understand why we roll a custom (de)serializer rather than building off the TOML crate?

**Comment by wmmc88** [comment]
> Detailed review is in progress, but quick design question: can you help me understand why we roll a custom (de)serializer rather than building off the TOML crate?

@NateD-MSFT the custom serde serializer takes the rust `metadata::Wdk` struct and turns it into a map that's eventually emitted as cargo prints to set `cfg's`. There is no similar "to_map" serializer.

There is no custom deserializer here. `cargo_metadata` [deserializes all the cargo metadata](https://github.com/wmmc88/windows-drivers-rs/blob/b95365c6a5e836f4e5b18da40f46089d1e4a073e/crates/wdk-build/src/lib.rs#L228) into `serde_json::Value`, and then we [deserialize that `serde_json::Value` into the `metadata::Wdk` struct using `serde_json`](https://github.com/wmmc88/windows-drivers-rs/blob/b95365c6a5e836f4e5b18da40f46089d1e4a073e/crates/wdk-build/src/metadata/mod.rs#L188).

File: crates/wdk-build/src/cargo_make.rs
Comment (wmmc88) [inline crates/wdk-build/src/cargo_make.rs]: This doesn't work for workspaces that have WDK configs, but when the current crate being processed is not marked as a driver

File: README.md
Comment (NateD-MSFT) [inline README.md]: Not directly related to this PR, but do we actually want to have language stating people can roll their own global allocators? For kernel mode I expect we want to use (a variant of) ExAllocatePool2 or bust.

File: crates/wdk-build/rust-driver-makefile.toml
Comment (NateD-MSFT) [inline crates/wdk-build/rust-driver-makefile.toml:98]: Just checking - is this task still necessary given https://github.com/sagiegurari/cargo-make/issues/1081 is now resolved and released?

File: crates/wdk-build/rust-driver-makefile.toml
Comment (NateD-MSFT) [inline crates/wdk-build/rust-driver-makefile.toml:438]:
```suggestion
env = { "WDK_BUILD_INF2CAT_OS" = { source = "${CARGO_MAKE_CRATE_TARGET_TRIPLE}", default_value = "UNKNOWN", mapping = { "x86_64-pc-windows-msvc" = "10_NI_X64", "aarch64-pc-windows-msvc" = "10_NI_ARM64" } } }
```

Comment (NateD-MSFT) [reply crates/wdk-build/rust-driver-makefile.toml:438]: I assume we want inf2cat to use the rules for the current version of Windows.

File: crates/wdk-build/src/lib.rs
Comment (NateD-MSFT) [inline crates/wdk-build/src/lib.rs:680]: Not something that needs to be changed in this PR, but in the future let's see if we can refactor all this target-specific logic out of raw code and into a declarative manner.  (Maybe I need to build a deserializer for the MSBuild proj files directly?)

File: crates/wdk-build/src/lib.rs
Comment (NateD-MSFT) [inline crates/wdk-build/src/lib.rs]:
```suggestion
/// Find the path of the toplevel Cargo manifest of the currently executing
```

File: crates/wdk-build/src/lib.rs
Comment (NateD-MSFT) [inline crates/wdk-build/src/lib.rs]:
```suggestion
/// lockfile. This does not support invocations that use non-default target
```

File: crates/wdk-build/src/lib.rs
Comment (NateD-MSFT) [inline crates/wdk-build/src/lib.rs:974]: Lazy is now stable if I'm not mistaken.  Not something that needs to be fixed right this instant though.

File: README.md
Comment (wmmc88) [reply README.md]: At the langauge level, rust has this bring-your-own-allocator construct for no_std programs (ex. km driver). There are still quite a few shortcomings with the `wdk-alloc` crate (ex. panic on oom) so for now I'd like to not force any sort of restrictions here. 

But given that the README is meant as a "getting started", I can change the wording to not explicitly encourage custom allocators

File: crates/wdk-build/rust-driver-makefile.toml
Comment (wmmc88) [reply crates/wdk-build/rust-driver-makefile.toml:98]: Nope. I just didn't want to make the change in this PR since scope is already extremely large here. Will address in a followup. Tracking this in https://github.com/microsoft/windows-drivers-rs/issues/171 already

File: crates/wdk-build/rust-driver-makefile.toml
Comment (wmmc88) [reply crates/wdk-build/rust-driver-makefile.toml:438]: I used `10_x64` as that was the default value that the WDK resolves to when building C drivers. 

Is there a new value every WDK release (I see `NI` in your suggestion)? If so, there'd need to be some logic here to resolve the mapping here based on the WDK config.

I'm inclined to leave it as is for this PR, since it matches parity with building C drivers

File: crates/wdk-build/src/lib.rs
Comment (wmmc88) [reply crates/wdk-build/src/lib.rs:680]: Yes. Ideally we'd be able to consume the WDK's various xml files (vcxproj, targets, etc) and get all of this information. There'd still need to be code here to do things to emit the correctly formatted print statements to pass the information to cargo, but all that information would be parsed from the xmls instead of hardcoded here.

> Maybe I need to build a deserializer for the MSBuild proj files directly

There would need to be work done to be able to merge all the xmls together, since many of the msbuild files reference other xml files and merge them together in some deterministic fashion. I'd imagine merging all that together, and then deserializing that merged xml using some serde xml deserializer(ex. serde_xml_rs) could work.

File: crates/wdk-build/src/lib.rs
Comment (wmmc88) [reply crates/wdk-build/src/lib.rs:974]: > Lazy is now stable if I'm not mistaken. Not something that needs to be fixed right this instant though.

Its available in stable as of 1.80.0 (released 5 days ago). Although not formally stated (or enforced) I've been trying to keep support for Stable and Stable-1. 

There are also other places in the repo where lazy_static is used, that aren't supported by std::sync::Lazy. (ex. LazyLock is not available in no_std, so in those situations, we should migrate to once_cell w/ the `racy` feature)

File: crates/wdk-build/rust-driver-makefile.toml
Comment (NateD-MSFT) [reply crates/wdk-build/rust-driver-makefile.toml:438]: When you say the default value, did you get that from WDK props, or...?

In any event see https://learn.microsoft.com/en-us/windows-hardware/drivers/devtest/inf2cat here for the full list of supported OS flags.  You can also provide multiple values to validate for multiple different OSes.  I think at the very least we should go to 10_NI for our default here since Rust drivers are at this early stage experimental and based on our latest OS planning.

File: README.md
Comment (NateD-MSFT) [reply README.md]: Yeah my concern is longer term - that we may have security or compliance reasons that we require people to use ExAllocatePool* and a specific MS allocator.  But your point about the current allocator being pretty bare bones is perfectly fair.  We can revisit this down the line.

File: crates/wdk-build/rust-driver-makefile.toml
Comment (wmmc88) [reply crates/wdk-build/rust-driver-makefile.toml:438]: This particular inf2cat setting corresponds with the `Inf2CatWindowsVersionList` property. 

On arm64, it defaults to `Server10_$(DDKPlatform)` where `DDKPlatform` resolves to `arm64`.
On x64, it defaults to `10_$(DDKPlatform)` where `DDKPlatform` resolves to `x64`.
In GE WDK, these defaults are set in `content\Program Files\Windows Kits\10\build\10.0.26100.0\WindowsDriver.OS.Props`.

If you open one of the sample windows drivers (ex. `umdf2toaster.sln`) in VS22, it confirms the above:
**x64:**
image
image
**arm64**
image
image

Comment (NateD-MSFT) [reply crates/wdk-build/rust-driver-makefile.toml:438]: For others reading this review: we resolved this offline and this is not blocking.

**Review by NateD-MSFT — APPROVED** [review]

File: crates/wdk-build/src/cargo_make.rs
Comment (wmmc88) [reply crates/wdk-build/src/cargo_make.rs]: fixed

File: README.md
Comment (wmmc88) [reply README.md]: I've updated the verbiage a bit to not specifically call out the functionality to bring in your own rust allocators

File: crates/wdk-build/src/lib.rs
Comment (wmmc88) [reply crates/wdk-build/src/lib.rs:974]: tracking this in https://github.com/microsoft/windows-drivers-rs/issues/192

**Comment by wmmc88** [comment]
Last round of changes for this PR fixed a logic bug that broke automatic skipping of the packaging task for non-binary(ie driver) wdk packages, added more tests for that specific situation, fixed various issues identified by pipelines(ie. formatting, clippy), and expanded coverage of the pipelines to include the workspace-level tests and examples folders.

File: crates/wdk-build/rust-driver-makefile.toml
Comment (wmmc88) [inline crates/wdk-build/rust-driver-makefile.toml]: The proper expansion of this is blocked on https://github.com/sagiegurari/cargo-make/pull/1132

Comment (wmmc88) [reply crates/wdk-build/rust-driver-makefile.toml]: added workaround to unblock pr

File: crates/wdk-build/rust-driver-makefile.toml
Comment (wmmc88) [inline crates/wdk-build/rust-driver-makefile.toml:137]: This should be removed once https://github.com/sagiegurari/cargo-make/pull/1132 is merged. The entirety of [2d199bc9e2b12e29f329fc9377ef31f498d35335 commit](https://github.com/microsoft/windows-drivers-rs/commit/2d199bc9e2b12e29f329fc9377ef31f498d35335) can be reverted after upstream release

Comment (wmmc88) [reply crates/wdk-build/rust-driver-makefile.toml:137]: tracked in #194

File: crates/wdk-build/src/metadata/error.rs
Comment (ayodejiige) [inline crates/wdk-build/src/metadata/error.rs:40]: what about if there are more than two values of the same key? is the intention to just catch the first duplicate?

File: crates/wdk-build/src/cargo_make.rs
Comment (ayodejiige) [inline crates/wdk-build/src/cargo_make.rs]: determines

**Review by ayodejiige — DISMISSED** [review]

File: crates/wdk-build/src/metadata/error.rs
Comment (wmmc88) [reply crates/wdk-build/src/metadata/error.rs:40]: It fails as soon as it finds the first duplicate. In the context of where this sits in the serializer apis and how those apis are invoked, I don't actually think its possible for a duplicate key to be output from the serializer. `-` is used as the seperator between rust identifiers, and rust identifiers aren't allowed to have `-` in them (even raw identifiers can't).

Comment (wmmc88) [reply crates/wdk-build/src/metadata/error.rs:40]: I had written this as an error instead of a panic because I actually wrote a deserializer as well that could have introduced this as a possible error condition. I ended up removing the deserializer from the scope of this PR, but plan to add it in a future followup

File: crates/wdk-build/src/cargo_make.rs
Comment (wmmc88) [reply crates/wdk-build/src/cargo_make.rs]: fixed

**Review by ayodejiige — APPROVED** [review]

**Review by hamzamust — APPROVED** [review]

**Pull request merged**
